pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -174,27 +209,93 @@ func (c *Controller) syncGKENetworkParamSet(ctx context.Context, key string) err
 	}
 
 	params := obj.(*networkv1.GKENetworkParamSet)
+	originalParams := params.DeepCopy()
 
-	// TODO: Enable finalizer addition when finalizer deletion is added.
-	// addFinalizerToGKENetworkParamSet(params)
-	// update will be done once in deferred function call
-	// if err := c.updateGKENetworkParamSet(ctx, params); err != nil {
-	// 	return err
-	// }
+	defer func() {
+		if !reflect.DeepEqual(originalParams, params) {
+			if updateErr := c.updateGKENetworkParamSet(ctx, params); updateErr != nil {
+				err = multierror.Append(updateErr, err)
+			}
+		}
+
+		if !reflect.DeepEqual(originalParams.Status, params.Status) {
+			if updateErr := c.updateGKENetworkParamSetStatus(ctx, params); updateErr != nil {
+				err = multierror.Append(updateErr, err)
+			}
+		}
+	}()
+
+	if params.DeletionTimestamp != nil {
+		// GKENetworkParamSet is being deleted, handle the delete event
+		return c.handleGKENetworkParamSetDelete(ctx, params)
+	}
 
-	subnet, err := c.gceCloud.GetSubnetwork(c.gceCloud.Region(), params.Spec.VPCSubnet)
+	addFinalizerInPlace(params)
+
+	subnet, subnetValidation := c.getAndValidateSubnet(ctx, params)
+	params.Status.Conditions = updateConditions(params.Status.Conditions, subnetValidation.toCondition())
+	if !subnetValidation.IsValid {
+		return
+	}
+
+	paramsValidation, err := c.validateGKENetworkParamSet(ctx, params, subnet)
 	if err != nil {
-		fetchSubnetErrs.Inc()
-		return err
+		return
+	}
+	params.Status.Conditions = updateConditions(params.Status.Conditions, paramsValidation.toCondition())
+	if !paramsValidation.IsValid {
+		return
 	}
 
 	cidrs := extractRelevantCidrs(subnet, params)
+	params.Status.PodCIDRs = &networkv1.NetworkRanges{
+		CIDRBlocks: cidrs,
+	}
 
-	err = c.updateGKENetworkParamSetStatus(ctx, params, cidrs)
+	networks, err := c.networkClientset.NetworkingV1().Networks().List(ctx, v1.ListOptions{})
 	if err != nil {
-		return err
+		return
+	}
+	// see if one of the networks is referencing this GNP
+	for _, network := range networks.Items {
+		if network.Spec.ParametersRef.Name == params.Name && network.Spec.ParametersRef.Kind == gnpKind {
+			c.syncNetworkWithGNP(ctx, &network, params)
+			break
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) syncNetworkWithGNP(ctx context.Context, network *networkv1.Network, params *networkv1.GKENetworkParamSet) (err error) {
+	newNetwork := network.DeepCopy()
+
+	defer func() {
+		if reflect.DeepEqual(newNetwork, network) {
```

Comment:
don't rely on reflect.DeepEquals, those are very expensive operations, you just know at the end of the function if you need to change the status or not

Reply:
for updating network you're right, the defer is overkill

Reply:
removed the defer here